File: tools/testing/connection/httpx/confluence.py
#!/usr/bin/env python3
"""
Confluence API - httpx Connection Test

Authentication: Basic (email:api_token)
Base URL: https://{company}.atlassian.net
Health Endpoint: GET /wiki/rest/api/user/current

TLS/SSL Options:
  SSL_CERT_VERIFY=0              - Ignore all certificate errors
  REQUEST_CA_BUNDLE=/path/to/ca  - Custom CA bundle file
  SSL_CERT_FILE=/path/to/cert    - Custom SSL certificate file
  REQUESTS_CA_BUNDLE=/path/to/ca - Alternative CA bundle (requests compat)
"""

import asyncio
import base64
import json
import os
from typing import Any, Union

import httpx

# ============================================================================
# Configuration - Override these values
# ============================================================================

CONFIG = {
    # Required (falls back to Jira credentials)
    "CONFLUENCE_BASE_URL": os.getenv("CONFLUENCE_BASE_URL", os.getenv("JIRA_BASE_URL", "https://your-company.atlassian.net")),
    "CONFLUENCE_EMAIL": os.getenv("CONFLUENCE_EMAIL", os.getenv("JIRA_EMAIL", "your.email@example.com")),
    "CONFLUENCE_API_TOKEN": os.getenv("CONFLUENCE_API_TOKEN", os.getenv("JIRA_API_TOKEN", "your_api_token_here")),

    # Optional: Proxy Configuration
    "HTTPS_PROXY": os.getenv("HTTPS_PROXY", ""),  # e.g., "http://proxy.example.com:8080"

    # Optional: TLS Configuration
    # Set SSL_CERT_VERIFY=0 to ignore certificate errors
    "VERIFY_SSL": os.getenv("SSL_CERT_VERIFY", "1") != "0",

    # Optional: Custom CA certificates
    # REQUEST_CA_BUNDLE, SSL_CERT_FILE, or REQUESTS_CA_BUNDLE - path to custom CA bundle
    "CA_BUNDLE": os.getenv("REQUEST_CA_BUNDLE") or os.getenv("SSL_CERT_FILE") or os.getenv("REQUESTS_CA_BUNDLE") or "",
}


# ============================================================================
# TLS Configuration Helper
# ============================================================================

def get_ssl_verify() -> Union[bool, str]:
    """Get SSL verification setting."""
    if CONFIG["CA_BUNDLE"]:
        print(f"Using custom CA bundle: {CONFIG['CA_BUNDLE']}")
        return CONFIG["CA_BUNDLE"]
    return CONFIG["VERIFY_SSL"]


# ============================================================================
# Create HTTP Client
# ============================================================================

def create_client() -> httpx.AsyncClient:
    """Create httpx async client with optional proxy."""
    proxies = None
    if CONFIG["HTTPS_PROXY"]:
        print(f"Using proxy: {CONFIG['HTTPS_PROXY']}")
        proxies = {
            "http://": CONFIG["HTTPS_PROXY"],
            "https://": CONFIG["HTTPS_PROXY"],
        }

    return httpx.AsyncClient(
        proxies=proxies,
        verify=get_ssl_verify(),
        timeout=httpx.Timeout(30.0),
    )


def create_basic_auth_header() -> str:
    """Create Basic auth header."""
    credentials = f"{CONFIG['CONFLUENCE_EMAIL']}:{CONFIG['CONFLUENCE_API_TOKEN']}"
    encoded = base64.b64encode(credentials.encode()).decode()
    return f"Basic {encoded}"


# ============================================================================
# Health Check
# ============================================================================

async def health_check() -> dict[str, Any]:
    """Perform health check against Confluence API."""
    print("\n=== Confluence Health Check ===\n")

    async with create_client() as client:
        try:
            response = await client.get(
                f"{CONFIG['CONFLUENCE_BASE_URL']}/wiki/rest/api/user/current",
                headers={
                    "Authorization": create_basic_auth_header(),
                    "Accept": "application/json",
                    "Content-Type": "application/json",
                },
            )

            data = response.json()

            print(f"Status: {response.status_code}")
            print(f"Response: {json.dumps(data, indent=2)}")

            return {"success": response.status_code == 200, "data": data}
        except Exception as e:
            print(f"Error: {e}")
            return {"success": False, "error": str(e)}


# ============================================================================
# Sample API Calls
# ============================================================================

async def list_spaces() -> dict[str, Any]:
    """List all spaces."""
    print("\n=== List Spaces ===\n")

    async with create_client() as client:
        try:
            response = await client.get(
                f"{CONFIG['CONFLUENCE_BASE_URL']}/wiki/rest/api/space",
                headers={
                    "Authorization": create_basic_auth_header(),
                    "Accept": "application/json",
                },
            )

            data = response.json()

            print(f"Status: {response.status_code}")
            results = data.get("results", [])
            print(f"Found {len(results)} spaces")
            for space in results[:10]:
                print(f"  - {space['key']}: {space['name']}")

            return {"success": response.status_code == 200, "data": data}
        except Exception as e:
            print(f"Error: {e}")
            return {"success": False, "error": str(e)}


async def search_content(cql: str) -> dict[str, Any]:
    """Search content using CQL."""
    print(f"\n=== Search Content: {cql} ===\n")

    async with create_client() as client:
        try:
            response = await client.get(
                f"{CONFIG['CONFLUENCE_BASE_URL']}/wiki/rest/api/content/search",
                params={"cql": cql, "limit": 10},
                headers={
                    "Authorization": create_basic_auth_header(),
                    "Accept": "application/json",
                },
            )

            data = response.json()

            print(f"Status: {response.status_code}")
            results = data.get("results", [])
            print(f"Found {len(results)} results")
            for content in results[:5]:
                print(f"  - {content['title']} ({content['type']})")

            return {"success": response.status_code == 200, "data": data}
        except Exception as e:
            print(f"Error: {e}")
            return {"success": False, "error": str(e)}


async def get_page(page_id: str) -> dict[str, Any]:
    """Get page details."""
    print(f"\n=== Get Page: {page_id} ===\n")

    async with create_client() as client:
        try:
            response = await client.get(
                f"{CONFIG['CONFLUENCE_BASE_URL']}/wiki/rest/api/content/{page_id}",
                params={"expand": "body.storage,version"},
                headers={
                    "Authorization": create_basic_auth_header(),
                    "Accept": "application/json",
                },
            )

            data = response.json()

            print(f"Status: {response.status_code}")
            print(f"Response: {json.dumps(data, indent=2)}")

            return {"success": response.status_code == 200, "data": data}
        except Exception as e:
            print(f"Error: {e}")
            return {"success": False, "error": str(e)}


# ============================================================================
# Run Tests
# ============================================================================

async def main():
    """Run connection tests."""
    print("Confluence API Connection Test")
    print("==============================")
    print(f"Base URL: {CONFIG['CONFLUENCE_BASE_URL']}")
    print(f"Proxy: {CONFIG['HTTPS_PROXY'] or 'None'}")
    print(f"Email: {CONFIG['CONFLUENCE_EMAIL']}")
    print(f"SSL Verify: {CONFIG['VERIFY_SSL']}")
    print(f"CA Bundle: {CONFIG['CA_BUNDLE'] or 'System default'}")

    await health_check()
    # await list_spaces()
    # await search_content("type=page")
    # await get_page("123456")


if __name__ == "__main__":
    asyncio.run(main())

File: tools/testing/connection/undici/confluence.mjs
/**
 * Confluence API - undici Connection Test
 *
 * Authentication: Basic (email:api_token)
 * Base URL: https://{company}.atlassian.net
 * Health Endpoint: GET /wiki/rest/api/user/current
 *
 * TLS/SSL Options:
 *   NODE_TLS_REJECT_UNAUTHORIZED=0  - Ignore all certificate errors
 *   REQUEST_CA_BUNDLE=/path/to/ca   - Custom CA bundle file
 *   SSL_CERT_FILE=/path/to/cert     - Custom SSL certificate file
 *   NODE_EXTRA_CA_CERTS=/path/to/ca - Additional CA certificates
 */

import { request, ProxyAgent, Agent } from 'undici';
import { readFileSync } from 'node:fs';

// ============================================================================
// Configuration - Override these values
// ============================================================================

const CONFIG = {
  // Required
  CONFLUENCE_BASE_URL: process.env.CONFLUENCE_BASE_URL || process.env.JIRA_BASE_URL || 'https://your-company.atlassian.net',
  CONFLUENCE_EMAIL: process.env.CONFLUENCE_EMAIL || process.env.JIRA_EMAIL || 'your.email@example.com',
  CONFLUENCE_API_TOKEN: process.env.CONFLUENCE_API_TOKEN || process.env.JIRA_API_TOKEN || 'your_api_token_here',

  // Optional: Proxy Configuration
  HTTPS_PROXY: process.env.HTTPS_PROXY || '', // e.g., 'http://proxy.example.com:8080'

  // Optional: TLS Configuration
  // Set to false to ignore certificate errors (default: false for testing)
  // NODE_TLS_REJECT_UNAUTHORIZED=0, REQUEST_CA_BUNDLE=null, SSL_CERT_FILE=null, NODE_EXTRA_CA_CERTS=null
  REJECT_UNAUTHORIZED: false,

  // Optional: Custom CA certificates (disabled by default for testing)
  CA_BUNDLE: null,
};

// ============================================================================
// TLS Configuration Helper
// ============================================================================

function getTlsConnectOptions() {
  const options = {
    rejectUnauthorized: CONFIG.REJECT_UNAUTHORIZED,
  };

  // Load custom CA bundle if specified
  if (CONFIG.CA_BUNDLE) {
    try {
      options.ca = readFileSync(CONFIG.CA_BUNDLE);
      console.log(`Using custom CA bundle: ${CONFIG.CA_BUNDLE}`);
    } catch (err) {
      console.warn(`Warning: Could not read CA bundle: ${err.message}`);
    }
  }

  return options;
}

// ============================================================================
// Create Dispatcher (with or without proxy)
// ============================================================================

function createDispatcher() {
  const tlsOptions = getTlsConnectOptions();

  if (CONFIG.HTTPS_PROXY) {
    console.log(`Using proxy: ${CONFIG.HTTPS_PROXY}`);
    return new ProxyAgent({
      uri: CONFIG.HTTPS_PROXY,
      connect: tlsOptions,
    });
  }
  return new Agent({
    connect: tlsOptions,
  });
}

// ============================================================================
// Create Basic Auth Header
// ============================================================================

function createBasicAuthHeader() {
  const credentials = `${CONFIG.CONFLUENCE_EMAIL}:${CONFIG.CONFLUENCE_API_TOKEN}`;
  return `Basic ${Buffer.from(credentials).toString('base64')}`;
}

// ============================================================================
// Health Check
// ============================================================================

async function healthCheck() {
  console.log('\n=== Confluence Health Check ===\n');

  const dispatcher = createDispatcher();

  try {
    const response = await request(`${CONFIG.CONFLUENCE_BASE_URL}/wiki/rest/api/user/current`, {
      method: 'GET',
      headers: {
        'Authorization': createBasicAuthHeader(),
        'Accept': 'application/json',
        'Content-Type': 'application/json',
      },
      dispatcher,
    });

    const data = await response.body.json();

    console.log(`Status: ${response.statusCode}`);
    console.log('Response:', JSON.stringify(data, null, 2));

    return { success: response.statusCode === 200, data };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await dispatcher.close();
  }
}

// ============================================================================
// Sample API Calls
// ============================================================================

async function listSpaces() {
  console.log('\n=== List Spaces ===\n');

  const dispatcher = createDispatcher();

  try {
    const response = await request(`${CONFIG.CONFLUENCE_BASE_URL}/wiki/rest/api/space`, {
      method: 'GET',
      headers: {
        'Authorization': createBasicAuthHeader(),
        'Accept': 'application/json',
      },
      dispatcher,
    });

    const data = await response.body.json();

    console.log(`Status: ${response.statusCode}`);
    console.log(`Found ${data.results?.length || 0} spaces`);
    data.results?.slice(0, 10).forEach((space) => {
      console.log(`  - ${space.key}: ${space.name}`);
    });

    return { success: response.statusCode === 200, data };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await dispatcher.close();
  }
}

async function searchContent(cql) {
  console.log(`\n=== Search Content: ${cql} ===\n`);

  const dispatcher = createDispatcher();

  try {
    const url = new URL(`${CONFIG.CONFLUENCE_BASE_URL}/wiki/rest/api/content/search`);
    url.searchParams.set('cql', cql);
    url.searchParams.set('limit', '10');

    const response = await request(url.toString(), {
      method: 'GET',
      headers: {
        'Authorization': createBasicAuthHeader(),
        'Accept': 'application/json',
      },
      dispatcher,
    });

    const data = await response.body.json();

    console.log(`Status: ${response.statusCode}`);
    console.log(`Found ${data.results?.length || 0} results`);
    data.results?.slice(0, 5).forEach((content) => {
      console.log(`  - ${content.title} (${content.type})`);
    });

    return { success: response.statusCode === 200, data };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await dispatcher.close();
  }
}

async function getPage(pageId) {
  console.log(`\n=== Get Page: ${pageId} ===\n`);

  const dispatcher = createDispatcher();

  try {
    const url = new URL(`${CONFIG.CONFLUENCE_BASE_URL}/wiki/rest/api/content/${pageId}`);
    url.searchParams.set('expand', 'body.storage,version');

    const response = await request(url.toString(), {
      method: 'GET',
      headers: {
        'Authorization': createBasicAuthHeader(),
        'Accept': 'application/json',
      },
      dispatcher,
    });

    const data = await response.body.json();

    console.log(`Status: ${response.statusCode}`);
    console.log('Response:', JSON.stringify(data, null, 2));

    return { success: response.statusCode === 200, data };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await dispatcher.close();
  }
}

// ============================================================================
// Run Tests
// ============================================================================

async function main() {
  console.log('Confluence API Connection Test');
  console.log('==============================');
  console.log(`Base URL: ${CONFIG.CONFLUENCE_BASE_URL}`);
  console.log(`Proxy: ${CONFIG.HTTPS_PROXY || 'None'}`);
  console.log(`Email: ${CONFIG.CONFLUENCE_EMAIL}`);
  console.log(`TLS Verify: ${CONFIG.REJECT_UNAUTHORIZED}`);
  console.log(`CA Bundle: ${CONFIG.CA_BUNDLE || 'System default'}`);

  await healthCheck();
  // await listSpaces();
  // await searchContent('type=page');
  // await getPage('123456');
}

main().catch(console.error);
